pkg/controller/unreachable/unreachable_controller.go

```diff
@@ -169,29 +169,35 @@ func (r *ReconcileRemoteMachineSet) Reconcile(request reconcile.Request) (reconc
 	remoteClientBuilder := r.remoteClusterAPIClientBuilder(cd)
 	var unreachableError error
 	updateUnreachable := true
-	// Attempt to connect to the remote cluster using the preferred API URL.
-	_, primaryErr := remoteClientBuilder.UsePrimaryAPIURL().Build()
-	if primaryErr != nil {
-		// If the remote cluster is not accessible via the preferred API URL, check if there is a fallback API URL to use.
-		if hasOverride(cd) {
-			cdLog.WithError(primaryErr).Info("unable to create remote API client using API URL override")
-			// If a connectivity recheck is needed or the remote cluster was reachable via the preferred API URL prior
-			// to this reconciliation, then attempt to connect to the remote cluster using the fallback API URL.
-			// Even when the controller continues to reconcile a ClusterDeployment waiting for the preferred API URL to
-			// become accessible, the controller should not recheck connectivity via the fallback API URL more often
-			// than once every 2 hours.
-			if connectivityRecheckNeeded || wasPrimaryActive {
-				if _, secondaryErr := remoteClientBuilder.UseSecondaryAPIURL().Build(); secondaryErr != nil {
-					cdLog.WithError(secondaryErr).Warn("unable to create remote API client with either the initial API URL or the API URL override, marking cluster unreachable")
-					unreachableError = utilerrors.NewAggregate([]error{primaryErr, secondaryErr})
+	var primaryErr error
+	fakeCluster := controllerutils.IsFakeCluster(cd)
```

Comment:
Couldn't see this being used outside of the if condition

Reply:
> Looking at all the places that now have conditional blocks I was thinking if we could replace these with some interface and its implementations?
> 
> InstallPodManager => actual impl , faked out ...
> I’m just thinking because there is quite a bit of actual code is now in this conditions.. which feels very weird to me..

This would be nice, I've toyed with this idea for places we fork based on cloud provider in the past as well though it wasn't super popular at the time.

The install manager part is tough, I want a huge portion of the normal install manager to run, and just skip on some very small pieces. That one in particular would be tough to push out to a fake interface. We skip/fake upload of admin pw, cluster ID, infra ID, and the execution of the create cluster command as well as log upload afterwards. I don't want to fake the entire installmanager for that. I guess we could use a little composition, one struct for uploading info, another for executing the create cluster command? WDYT, worth it?

Throughout the controllers we could potentially setup some kind of GetClusterAdapter(cd), and that adapter could expose something like SetClusterStatusURLs, GetClusterOperators, GetHelperForRestConfig, GetClusterVersion, GetMachineSets, Unreachable. Actually all of those boil down close to that remoteclient.ConnectToRemoteCluster function, perhaps that could check the annotation and return a fake client like we use in unit tests, with the various hard coded data we'd want to return for that cluster. That might work... 

> Also, if we do go about this then our fake and real implementation are 2 different things, being maintained separately... soare we really faking the real thing?

We definitely need to be very careful with that. In this PR alone we're losing that serialization hit, though we could probably restore it. 

> 
> Secondly, why do we want a per cd annotation to fake clusters?
> 
>     * I’m not sure what workflow would benefit from having real cds and faked cds together being acted upon
> 
>     * also this makes it so actual users can do this for their cds which I don’t think is appropriate..
>       Is is too bad to keep hive configuration that you in fake mode and fake out all you can?

The stage test workflow would benefit from this and is the primary driver for the work. We don't have fully dedicated clusters for stage testing and instead we run those tests in existing clusters, which do contain some active clusters. It makes it simpler for a test suite driving that test to be able to control the behavior for it's own clusterdeployments, without requiring an admin or CI change to flip global hiveconfig. The latter was ok for us on the dev side but there's a lot of interest in this on SD side to be able to try to replicate the load on etcd, and api, with OCM and SRE operators in the mix, without running up massive bills and having to rush to finish and clean up.

With regard to users being able to use this, this is why I go with annotations rather than anything documented and commited to in the actual API Spec. We do this in a couple places to let users simulate things that we don't necessarily want to draw attention to or commit to.